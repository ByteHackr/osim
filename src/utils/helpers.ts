import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';
import * as R from 'ramda';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );

type DeepMappable = any[] | Record<string, any>

const isNonEmptyArray = (value: any) => R.is(Array, value) && value.length > 0;
const isNonArrayObject = (value: any) => R.is(Object, value) && !R.is(Array, value);
const isDeepMappable = (value: DeepMappable) => isNonEmptyArray(value) || isNonArrayObject(value);

export const deepMap = (transform: (arg: any) => any, object: DeepMappable): any =>
  R.map(
    (val: any) => isDeepMappable(val)
      ? deepMap(transform, val)
      : transform(val)
    , object
  );

export const cveRegex = /^\s*(CVE-\d{4}-\d{4,7})\s*$/;
