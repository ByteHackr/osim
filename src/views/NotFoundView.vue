<script setup lang="ts">

</script>

<template>
  <main class="mt-3 container">
    <h1 class="mb-3">Page Not Found</h1>
    <RouterLink class="btn btn-danger" :to="{name: 'index'}">Back to Index</RouterLink>
  </main>
</template>
