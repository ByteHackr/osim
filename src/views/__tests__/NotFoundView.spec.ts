import { mountWithConfig } from '@/__tests__/helpers';

import NotFoundView from '../NotFoundView.vue';

describe('notFoundView', () => {
  it('should render', () => {
    const wrapper = mountWithConfig(NotFoundView);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
