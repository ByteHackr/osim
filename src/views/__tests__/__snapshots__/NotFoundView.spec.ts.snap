// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`notFoundView > should render 1`] = `
"<main class="mt-3 container">
  <h1 class="mb-3">Page Not Found</h1><a href="/" class="btn btn-danger">Back to Index</a>
</main>"
`;
